File: src/main/java/no/novari/flyt/configuration/model/configuration/dtos/CollectionMappingDto.java
package no.novari.flyt.configuration.model.configuration.dtos;

import jakarta.validation.Valid;
import jakarta.validation.constraints.NotNull;
import lombok.Builder;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.extern.jackson.Jacksonized;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Optional;

@Getter
@Builder
@EqualsAndHashCode
@Jacksonized
public class CollectionMappingDto<T> {

    public CollectionMappingDto(
            Collection<T> elementMappings,
            Collection<FromCollectionMappingDto<T>> fromCollectionMappings
    ) {
        this.elementMappings = Optional.ofNullable(elementMappings).orElse(new ArrayList<>());
        this.fromCollectionMappings = Optional.ofNullable(fromCollectionMappings).orElse(new ArrayList<>());
    }

    private final Collection<@Valid @NotNull T> elementMappings;

    private final Collection<@Valid @NotNull FromCollectionMappingDto<T>> fromCollectionMappings;

}

File: src/main/kotlin/no/novari/flyt/configuration/model/configuration/dtos/CollectionMappingDto.kt
package no.novari.flyt.configuration.model.configuration.dtos

import jakarta.validation.Valid
import jakarta.validation.constraints.NotNull

data class CollectionMappingDto<T>(
    @field:Valid
    var elementMappings: MutableCollection<@NotNull T> = mutableListOf(),
    @field:Valid
    var fromCollectionMappings: MutableCollection<@NotNull FromCollectionMappingDto<T>> = mutableListOf(),
) {
    companion object {
        @JvmStatic
        fun <T> builder(): Builder<T> = Builder()
    }

    class Builder<T> {
        private var elementMappings: MutableCollection<T> = mutableListOf()
        private var fromCollectionMappings: MutableCollection<FromCollectionMappingDto<T>> = mutableListOf()

        fun elementMappings(elementMappings: Collection<T>?) =
            apply { this.elementMappings = elementMappings?.toMutableList() ?: mutableListOf() }

        fun fromCollectionMappings(fromCollectionMappings: Collection<FromCollectionMappingDto<T>>?) =
            apply { this.fromCollectionMappings = fromCollectionMappings?.toMutableList() ?: mutableListOf() }

        fun build(): CollectionMappingDto<T> =
            CollectionMappingDto(
                elementMappings = elementMappings,
                fromCollectionMappings = fromCollectionMappings,
            )
    }
}
